pkg/tbtc/signing_loop.go

```diff
@@ -0,0 +1,478 @@
+package tbtc
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"math/rand"
+	"sort"
+
+	"github.com/ipfs/go-log/v2"
+	"github.com/keep-network/keep-core/pkg/chain"
+	"github.com/keep-network/keep-core/pkg/protocol/group"
+	"github.com/keep-network/keep-core/pkg/tecdsa/retry"
+	"github.com/keep-network/keep-core/pkg/tecdsa/signing"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	// signingAttemptAnnouncementDelayBlocks determines the duration of the
+	// announcement phase delay that is preserved before starting the
+	// announcement phase.
+	signingAttemptAnnouncementDelayBlocks = 1
+	// signingAttemptAnnouncementActiveBlocks determines the duration of the
+	// announcement phase that is performed at the beginning of each signing
+	// attempt.
+	signingAttemptAnnouncementActiveBlocks = 5
+	// signingAttemptProtocolBlocks determines the maximum block duration of the
+	// actual protocol computations.
+	signingAttemptMaximumProtocolBlocks = 20
+	// signingAttemptCoolDownBlocks determines the duration of the cool down
+	// period that is preserved between subsequent signing attempts.
+	signingAttemptCoolDownBlocks = 5
```

Comment:
The cooldown is only used in case of retries. This is the time we give for signers to recover and prepare for the next attempt and it is absolutely needed to make the retry loop work as expected. During the happy path, this cooldown is not applied and the signer returns as soon as it has the result.

Reply:
> (...) and it is absolutely needed to make the retry loop work as expected.

Why is it absolutely needed? What activities are performed as a part of "signer recovery" after the last failed attempt"?

Reply:
I think the explanation in the code may be helpful:
```
		// In order to start attempts >1 in the right place, we need to
		// determine how many blocks were taken by previous attempts. We assume
		// the worst case that each attempt failed at the end of the signing
		// protocol.
		//
		// That said, we need to increment the previous attempt start
		// block by the number of blocks equal to the protocol duration and
		// by some additional delay blocks. We need a small cool down in
		// order to mitigate all corner cases where the actual attempt duration
		// was slightly longer than the expected duration determined by the
		// signingAttemptMaximumProtocolBlocks constant.
		//
		// For example, the attempt may fail at the end of the protocol but the
		// error is returned after some time and more blocks than expected are
		// mined in the meantime.
		if srl.attemptCounter > 1 {
			srl.attemptStartBlock = srl.attemptStartBlock +
				uint64(signingAttemptMaximumBlocks())
		}
```

Reply:
I am sorry but I still do not understand it. 

The timeout passed to `signingAttemptFn` is computed as `announcementEndBlock + signingAttemptMaximumProtocolBlocks`. `signingAttemptMaximumProtocolBlocks` does include the `signingAttemptCoolDownBlocks`. 

Based on the third paragraph of the documentation you pasted, what I think is happening is that the cooldown matters only when the protocol is still executing just before the timeout. Then, the timeout is hit, the context of the attempt gets canceled, the loop logs a warning and we start the next attempt.

Given that the timeout includes `signingAttemptCoolDownBlocks` I do not see why this value is defined separately. It is only a part of `signingAttemptMaximumBlocks()` that is used to set the timeout for the context. Signers are not performing any cooldown activities, they keep trying to execute the protocol until `signingAttemptMaximumBlocks()` which includes `signingAttemptCoolDownBlocks`.

Reply:
What you probably missed is that `signingAttemptMaximumProtocolBlocks` DOES NOT include `signingAttemptCoolDownBlocks`. The `signingAttemptMaximumProtocolBlocks` value determines the maximum duration of the protocol within an attempt, not the entire attempt. The maximum duration of the whole attempt is `signingAttemptMaximumBlocks()` which DOES include `signingAttemptCoolDownBlocks`:
```
// signingAttemptMaximumBlocks returns the maximum block duration of a single
// signing attempt.
func signingAttemptMaximumBlocks() uint {
	return signingAttemptAnnouncementDelayBlocks +
		signingAttemptAnnouncementActiveBlocks +
		signingAttemptMaximumProtocolBlocks +
		signingAttemptCoolDownBlocks
}
```